### PR TITLE
[video] Add sampling capabilities

### PR DESCRIPTION
**Pull request by changhiskhan**
Create utilities for sampling videos using opencv then
integrate that into DataLoader to reduce boilerplate.
We use prefetch to sample videos in parallel but each video
is sampled by the same thread to keep the order of the frames.

### CODE REVIEW

File: python/rikai/torch/data.py [hunk 1]
```diff
@@ -131,3 +135,29 @@ def __iter__(self) -> Generator[Dict, None, None]:
             yield self.collate_fn(batch)
         prefetch_thd.join()
         q.join()
+
+
+def make_video_sampler(sample_gen, video_column_name=None):
```
Comment (eddyxu) [inline python/rikai/torch/data.py:140]: Add doc here? how should user expect to use it

File: python/rikai/video.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        :param max_samples: yield at most this many frames
+        """
+        super().__init__(stream)
+        self.sample_rate = sample_rate
+        self.start_frame = start_frame
+        self.max_samples = max_samples
+
+    def __iter__(self):
+        # TODO use seek for sparse sampling and maybe multithreaded
+        cap = cv2.VideoCapture(self.stream.uri)
```
Comment (eddyxu) [inline python/rikai/video.py:99]: do we need to put `cv2` into deps?

File: python/rikai/video.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+class SingleFrameSampler(VideoSampler):
+    """
+    A simple sampler that just returns one out of every `sample_rate` frames
+    """
+
+    def __init__(
+        self, stream: VideoStream, sample_rate: int, start_frame: int, max_samples: int
+    ):
+        """
+        :param stream: source video to sample from
```
Comment (eddyxu) [inline python/rikai/video.py:87]: We should have a style guide :) 

i used [numpy docstring](https://numpydoc.readthedocs.io/en/latest/format.html) for the rest of the python codebase.  Lets get some consent on docstring format ?

File: python/rikai/video.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
+from abc import ABC, abstractmethod
```
Comment (eddyxu) [inline python/rikai/video.py:1]: Apache license. 

And rename this to `python/rikai/types/video.py`?

File: python/rikai/video.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+import cv2
+import youtube_dl
+
+from rikai.spark.types import (
+    YouTubeVideoType,
+    VideoStreamType,
+)
+from rikai.vision import Image
```
Comment (eddyxu) [inline python/rikai/video.py:9]: Is this import needed?

File: python/rikai/video.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        from IPython.lib.display import YouTubeVideo
+
+        return YouTubeVideo(self.vid)._repr_html_()
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, YouTubeVideo) and self.vid == other.vid
+
+    def get_stream(self, **kwargs) -> "VideoStream":
+        # TODO allow configurable cache dir
+        dl_path = "/tmp/{}.mp4".format(self.vid)
```
Comment (eddyxu) [inline python/rikai/video.py:34]: use `TMPDIR`?

File: python/rikai/video.py (hunk 3 above)
Comment (changhiskhan) [reply python/rikai/video.py:87]: These were auto-generated by IntelliJ. I prefer numpy docstring actually, I'll do a docs/comments cleanup pass

File: python/rikai/video.py (hunk 4 above)
Comment (changhiskhan) [reply python/rikai/video.py:1]: Do you think the sampling classes should go in types or in like `python/rikai/utils.py` ?

File: python/rikai/video.py (hunk 5 above)
Comment (changhiskhan) [reply python/rikai/video.py:9]: copy-pasta. i'll remove.

File: python/rikai/torch/data.py (hunk 1 above)
Comment (changhiskhan) [reply python/rikai/torch/data.py:140]: I will add it there. Current usage is in the notebook i added. I can move some code snippets over.

File: python/rikai/video.py (hunk 3 above)
Comment (eddyxu) [reply python/rikai/video.py:87]: You can change intellij to set default docstring to numpy style.

File: python/rikai/video.py (hunk 4 above)
Comment (eddyxu) [reply python/rikai/video.py:1]: should we have a `python/rikai/video.py` for video related functionalities?

i'd imagine that we will have

```
rikai.types -> for UDT
rikai.torch -> for torch access
rikai.featurestore (or .dataset) -> for dataset management
rikai.video(.sample/.other function) -> for video specific operations?

```

you know what, what about `rikai.transforms.video`?

File: python/rikai/video.py (hunk 6 above)
Comment (changhiskhan) [reply python/rikai/video.py:34]: good idea. I will use TMPDIR for this PR. I need to make this more configurable and also automatically add an option to cp to s3 as well.

File: python/rikai/video.py (hunk 2 above)
Comment (changhiskhan) [reply python/rikai/video.py:99]: added

File: python/rikai/video.py (hunk 3 above)
Comment (changhiskhan) [reply python/rikai/video.py:87]: yeah, i had project settings set before but I created a new one for the new repo and lost it

File: python/rikai/video.py (hunk 4 above)
Comment (changhiskhan) [reply python/rikai/video.py:1]: i put everything in rikai.types.video for now. We can decide where to put these utility functions.

File: python/__init__.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#  Copyright 2021 Rikai Authors
```
Comment (eddyxu) [inline python/__init__.py:1]: Do we need this level ? 🤣

**Review by eddyxu — APPROVED** [review]
lgtm

File: python/__init__.py (hunk 7 above)
Comment (changhiskhan) [reply python/__init__.py:1]: I just have intellij automatically add it to all .py files lol

**Pull request merged**